server/util/cache_metrics/cache_metrics.go

```diff
@@ -0,0 +1,175 @@
+// which the cache sits in the composable cache stack.
+//
+// This label enables Prometheus queries that don't break when the cache implementation
+// is changed. For example, to get cache metrics for the in-memory tier only (without having
+// to explicitly refer to "redis", "memcached", etc.), one could write this query:
+//
+//     sum by (backend) (rate(buildbuddy_cache_write_count{tier="memory"}[5m])
+//
+// These types of queries are useful when monitoring the cache because caches
+// at different tiers will have observed latencies at different orders of magnitude,
+// so each tier likely needs its own chart in order to be observed at an appropriate
+// resolution.
+type CacheTier string
+
+const (
+	// MemoryCacheTier is used for in-memory cache implementations, like memcached and redis.
+	MemoryCacheTier CacheTier = "memory"
+	// CloudCacheTier is used for cloud storage cache implementations, like GCS or AWS S3.
+	CloudCacheTier CacheTier = "cloud"
+)
+
+func MakeCacheLabels(tier CacheTier, backend string) prometheus.Labels {
+	return prometheus.Labels{
+		metrics.CacheTierLabel:    string(tier),
+		metrics.CacheBackendLabel: backend,
+	}
+}
+
+func appendLabels(a prometheus.Labels, b prometheus.Labels) prometheus.Labels {
+	labels := prometheus.Labels{}
+	for k, v := range a {
+		labels[k] = v
+	}
+	for k, v := range b {
+		labels[k] = v
+	}
+	return labels
+}
+
+// InstrumentedReader wraps a Reader and CacheTimer so that the embedded timer is stopped
+// and metrics are recorded on the last read.
+type InstrumentedReader struct {
+	io.Reader
+	timer         *CacheTimer
+	readSizeBytes int64
+}
+
+func (ir *InstrumentedReader) Read(p []byte) (int, error) {
+	n, err := ir.Reader.Read(p)
+	// Record metrics on the last read.
+	if err != nil {
+		ir.timer.EndRead(ir.readSizeBytes, err)
+	}
+	return n, err
+}
+
+func RecordSetRetries(labels prometheus.Labels, numRetries int) {
+	metrics.CacheSetRetryCount.With(labels).Observe(float64(numRetries))
+}
+
+func RecordWriteRetries(labels prometheus.Labels, numRetries int) {
+	metrics.CacheWriteRetryCount.With(labels).Observe(float64(numRetries))
+}
+
+func RecordContainsRetries(labels prometheus.Labels, numRetries int) {
+	metrics.CacheContainsRetryCount.With(labels).Observe(float64(numRetries))
+}
+
+type CacheTimer struct {
+	labels    prometheus.Labels
+	startTime time.Time
+}
+
+func NewCacheTimer(labels prometheus.Labels) *CacheTimer {
+	return &CacheTimer{
+		labels:    labels,
+		startTime: time.Now(),
+	}
+}
+
+func (t *CacheTimer) NewInstrumentedReader(r io.Reader, size int64) *InstrumentedReader {
+	return &InstrumentedReader{
+		Reader:        r,
+		timer:         t,
+		readSizeBytes: size,
+	}
+}
+
+// Note: in all the following methods, we don't track duration if there's
+// an error, but do track count (recording the status code) so we can measure
+// failure rates.
+
+func (t *CacheTimer) EndGet(size int, err error) {
```

Comment:
I think your cachetimer is sort of close to something you'd see commonly in go, where you'd open some resource and defer a close. Like:
  f, err := os.Open(/path/to/file)
  ...
  defer f.Close()

But instead of recording your action type and size in the Open method, you do that in the EndAction method. This is OK, a little unintuitive, but I see why you did it and I think it keeps the code really clear. So let's keep that part!

But, one thing I think we could do to make this feel more natural, is to rename the methods from EndAction to something like.... RecordAction? LogAction? TimeAction? ObserveAction?

Reply:
I was a little unsatisfied with `End` myself, I like the `Observe` suggestion since it matches up with the Prometheus verb. Done!